sdk/helper/keysutil/policy.go

```diff
@@ -1596,3 +1514,136 @@ func (p *Policy) getVersionPrefix(ver int) string {
 
 	return prefix
 }
+
+// SymmetricOpts are the arguments to symmetric operations that are "optional", e.g.
+// not always used.  This improves the aesthetics of calls to those functions.
+type SymmetricOpts struct {
+	// Whether to use convergent encryption
+	Convergent bool
+	// The version of the convergent encryption scheme
+	ConvergentVersion int
+	// The nonce, if not randomly generated
+	Nonce []byte
+	// Additional data to include in AEAD authentication
+	AdditionalData []byte
+	// The HMAC key, for generating IVs in convergent encryption
+	HMACKey []byte
+}
+
+// Symmetrically encrypt a plaintext given the convergence configuration and appropriate keys
+func (p *Policy) SymmetricEncryptRaw(ver int, encKey, plaintext []byte, opts SymmetricOpts) ([]byte, error) {
+	var aead cipher.AEAD
+	var err error
+	nonce := opts.Nonce
+
+	switch p.Type {
+	case KeyType_AES128_GCM96, KeyType_AES256_GCM96:
+		// Setup the cipher
+		aesCipher, err := aes.NewCipher(encKey)
+		if err != nil {
+			return nil, errutil.InternalError{Err: err.Error()}
+		}
+
+		// Setup the GCM AEAD
+		gcm, err := cipher.NewGCM(aesCipher)
+		if err != nil {
+			return nil, errutil.InternalError{Err: err.Error()}
+		}
+
+		aead = gcm
+
+	case KeyType_ChaCha20_Poly1305:
+		cha, err := chacha20poly1305.New(encKey)
+		if err != nil {
+			return nil, errutil.InternalError{Err: err.Error()}
+		}
+
+		aead = cha
+	}
+
+	if opts.Convergent {
+		convergentVersion := p.convergentVersion(ver)
+		switch convergentVersion {
+		case 1:
+			if len(opts.Nonce) != aead.NonceSize() {
+				return nil, errutil.UserError{Err: fmt.Sprintf("base64-decoded nonce must be %d bytes long when using convergent encryption with this key", aead.NonceSize())}
+			}
+		case 2, 3:
+			if len(opts.HMACKey) == 0 {
+				return nil, errutil.InternalError{Err: fmt.Sprintf("invalid hmac key length of zero")}
+			}
+			nonceHmac := hmac.New(sha256.New, opts.HMACKey)
+			nonceHmac.Write(plaintext)
+			nonceSum := nonceHmac.Sum(nil)
+			nonce = nonceSum[:aead.NonceSize()]
+		default:
+			return nil, errutil.InternalError{Err: fmt.Sprintf("unhandled convergent version %d", convergentVersion)}
+		}
+	} else if nonce == nil {
```

Comment:
Can we do a length check here instead? Just to be sure we aren't being tricked by typed nils?

Reply:
Yep.